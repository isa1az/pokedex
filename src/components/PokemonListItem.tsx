import React from 'react';
import {Pokemon} from '../interfaces/pokemonResponse';
import {
  Dimensions,
  Image,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';

const windowWidth = Dimensions.get('window').width;

interface Props {
  pokemon: Pokemon;
}

const PokemonListItem = ({pokemon}: Props) => {
  const navigation = useNavigation();

  return (
    <TouchableOpacity
      activeOpacity={0.9}
      onPress={() => navigation.navigate('PokemonDetailScreen', {pokemon})}>
      <View style={[styles.itemContainer, {width: windowWidth * 0.275}]}>
        <Image source={{uri: pokemon.picture}} style={styles.pokemonImage} />

        <Text style={styles.name}>{pokemon.name}</Text>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  itemContainer: {
    backgroundColor: 'white',
    marginHorizontal: 10,
    height: 100,
    width: 120,
    marginBottom: 25,
    justifyContent: 'center',
    alignItems: 'center',
  },
  name: {
    textTransform: 'capitalize',
  },
  pokemonImage: {
    width: 70,
    height: 70,
  },
});

export default PokemonListItem;
