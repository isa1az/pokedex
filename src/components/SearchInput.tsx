import React, {useMemo, useState} from 'react';
import {Platform, StyleSheet, View, StyleProp, ViewStyle} from 'react-native';
import {TextInput} from 'react-native-gesture-handler';
import {Ionicons} from '@expo/vector-icons';
import {debounce} from 'lodash';

interface Props {
  onSearch: (value: string) => void;
  style?: StyleProp<ViewStyle>;
  disabled?: boolean;
}

export const SearchInput = ({style, onSearch, disabled = false}: Props) => {
  const [textValue, setTextValue] = useState('');

  const handleChangeWithDebounce = useMemo(
    () => debounce((text: string) => onSearch(text), 350),
    [onSearch],
  );

  return (
    <View
      style={{
        ...styles.container,
        ...(style as any),
      }}>
      <View style={styles.textBackground}>
        <TextInput
          placeholder="Buscar pokÃ©mon"
          style={{
            ...styles.textInput,
            top: Platform.OS === 'ios' ? 0 : 2,
          }}
          autoCapitalize="none"
          autoCorrect={false}
          value={textValue}
          onChangeText={value => {
            setTextValue(value);
            handleChangeWithDebounce(value);
          }}
          editable={!disabled}
          selectTextOnFocus={!disabled}
        />

        <Ionicons name="search-outline" color="grey" size={30} />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {},
  textBackground: {
    backgroundColor: '#F3F1F3',
    borderRadius: 50,
    height: 40,
    paddingHorizontal: 20,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,

    elevation: 5,
  },
  textInput: {
    flex: 1,
    fontSize: 18,
  },
});
