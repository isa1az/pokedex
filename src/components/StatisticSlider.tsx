import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {Stat} from '../interfaces/pokemonResponse';
import {parseStatName} from '../utils/utils';

interface Props {
  stat: Stat;
  maxStat: number;
}

const StatisticSlider = ({stat, maxStat}: Props) => {
  const statName = parseStatName(stat.stat.name);

  return (
    <View style={styles.container}>
      <Text style={styles.name}>{statName}</Text>

      <View style={styles.barContainer}>
        <View
          style={{
            ...styles.statContainer,
            width: `${(stat.base_stat * 100) / maxStat}%`,
          }}
        />
        <View style={styles.thumbContainer}>
          <Text style={styles.thumb}>{stat.base_stat}</Text>
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: 5,
  },
  name: {
    color: 'black',
    textTransform: 'capitalize',
    width: 70,
    marginRight: 5,
  },
  barContainer: {
    backgroundColor: 'lightgrey',
    height: 10,
    flex: 1,
    borderRadius: 3,
    flexDirection: 'row',
  },
  statContainer: {
    height: 10,
    backgroundColor: '#3d7dca',
    borderRadius: 3,
  },
  thumbContainer: {
    height: 10,
    flex: 1,
    justifyContent: 'center',
  },
  thumb: {
    backgroundColor: 'white',
    position: 'absolute',
    paddingVertical: 1,
    paddingHorizontal: 10,
    borderRadius: 3,
    left: -30,

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,

    elevation: 5,
  },
});

export default StatisticSlider;
