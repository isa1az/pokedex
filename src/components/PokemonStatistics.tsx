import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {PokemonFull} from '../interfaces/pokemonResponse';
import StatisticSlider from './StatisticSlider';
import {getMaxStat} from '../utils/utils';

interface Props {
  pokemon: PokemonFull;
  isLoading: boolean;
}

const PokemonStatistics = ({pokemon: {stats}}: Props) => {
  const maxStat = getMaxStat(stats);

  return (
    <View style={styles.container}>
      <View style={styles.titleContainer}>
        <View style={styles.titleBorder} />
        <Text style={styles.title}>Statistics</Text>
        <View style={styles.titleBorder} />
      </View>

      {stats?.map(stat => (
        <StatisticSlider
          key={`stat-${stat.stat.name}`}
          stat={stat}
          maxStat={maxStat}
        />
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    marginTop: 35,
  },
  titleContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 20,
  },
  titleBorder: {
    height: 1,
    width: 20,
    backgroundColor: 'black',
    flex: 2,
  },
  title: {
    textTransform: 'uppercase',
    marginHorizontal: 10,
    fontWeight: '600',
  },
});

export default PokemonStatistics;
