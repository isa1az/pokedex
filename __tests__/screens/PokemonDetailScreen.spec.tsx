import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import {SearchInput} from '../../src/components/SearchInput';

describe('<PokemonDetailScreen />', () => {
  it('should match snapshot', async () => {
    const onSearch = jest.fn();
    const wrapper = render(<SearchInput onSearch={onSearch} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger onSearch function after debounce', async () => {
    const onSearch = jest.fn();
    const {getByPlaceholderText} = render(<SearchInput onSearch={onSearch} />);

    const input = getByPlaceholderText('Search');

    fireEvent.changeText(input, 'Charmander');

    expect(onSearch).toHaveBeenCalledTimes(0);

    setTimeout(() => {
      expect(onSearch).toHaveBeenCalled();
    }, 350);
  });

  it('should send written text', async () => {
    const onSearch = jest.fn();
    const {getByPlaceholderText} = render(<SearchInput onSearch={onSearch} />);

    const input = getByPlaceholderText('Search');

    fireEvent.changeText(input, 'Charmander');

    setTimeout(() => {
      expect(onSearch).toHaveBeenCalledWith('Charmander');
    }, 350);
  });
});
