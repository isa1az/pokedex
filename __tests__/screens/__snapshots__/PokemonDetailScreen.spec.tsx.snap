// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonDetailScreen /> should match snapshot 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F3F1F3",
        "borderRadius": 50,
        "elevation": 5,
        "flexDirection": "row",
        "height": 40,
        "justifyContent": "center",
        "paddingHorizontal": 20,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      collapsable={false}
      editable={true}
      onChangeText={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      placeholder="Search"
      rejectResponderTermination={true}
      selectTextOnFocus={true}
      style={
        Object {
          "flex": 1,
          "fontSize": 18,
          "top": 0,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <Text />
  </View>
</View>
`;

exports[`<PokemonDetailScreen /> should trigger onSearch function after debounce 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F3F1F3",
        "borderRadius": 50,
        "elevation": 5,
        "flexDirection": "row",
        "height": 40,
        "justifyContent": "center",
        "paddingHorizontal": 20,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      collapsable={false}
      editable={true}
      onChangeText={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      placeholder="Search"
      rejectResponderTermination={true}
      selectTextOnFocus={true}
      style={
        Object {
          "flex": 1,
          "fontSize": 18,
          "top": 0,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <Text />
  </View>
</View>
`;
